ksql-rest-app/src/main/java/io/confluent/ksql/api/server/ServerVerticle.java

```diff
@@ -86,6 +109,107 @@ private Router setupRouter() {
         .handler(new InsertsStreamHandler(context, endpoints, server.getWorkerExecutor()));
     router.route(HttpMethod.POST, "/close-query").handler(BodyHandler.create())
         .handler(new CloseQueryHandler(server));
+    router.route(HttpMethod.GET, "/ws/*").handler(this::websocketHandler);
+
+    if (proxyEnabled) {
+      // Everything else is proxied
+      // The proxying is temporary and will go away once all the Jetty endpoints have been
+      // migrated to Vert.x
+      router.route().handler(new ProxyHandler(proxyTarget, proxyClient, server));
+    }
+    router.route().failureHandler(this::handleFailure);
     return router;
   }
+
+  private void handleFailure(final RoutingContext routingContext) {
+    if (routingContext.statusCode() == 500) {
+      log.error("Unexpected exception in router", routingContext.failure());
+    }
+    routingContext.response().setStatusCode(routingContext.statusCode()).end();
+  }
+
+  private void websocketHandler(final RoutingContext routingContext) {
+    if (proxyTarget == null) {
+      proxyTarget = server.getProxyTarget();
+    }
+    final HttpServerRequest request = routingContext.request();
+    final WebSocketConnectOptions options = new WebSocketConnectOptions()
+        .setHost(proxyTarget.host())
+        .setPort(proxyTarget.port())
+        .setHeaders(request.headers())
+        .setURI(request.uri());
+    request.pause();
+    proxyClient.webSocket(options, ar -> {
+      request.resume();
+      if (ar.succeeded()) {
+        final WebSocket webSocket = ar.result();
+        final ServerWebSocket serverWebSocket = request.upgrade();
+        WebsocketPipe.pipe(serverWebSocket, webSocket);
+        WebsocketPipe.pipe(webSocket, serverWebSocket);
+      } else {
+        if (ar.cause() instanceof UpgradeRejectedException) {
+          final UpgradeRejectedException uge = (UpgradeRejectedException) ar.cause();
```

Comment:
nit: should we add a log message here?

Reply:
This will occur for things like authorization failure, in which case presumably the old Jetty server already logs this.